docs/proposals/memory-management.md

```diff
@@ -0,0 +1,165 @@
+
+2. Apply memory limits to user containers and individual processes on the node.
+
+3. Set up cgroups to track the memory consumption of itself and other system processes like docker daemon, kube-proxy, sshd, etc. The cgroups being setup by kubelet are,
+  a. `/docker-daemon`: The docker daemon is under this cgroup. This cgroup is limited to 70% of the total available memory on the node.
+  b. `/system`: OS processes, excepting init, are under this cgroup.
+  c. `/kubelet`: Kubelet is under this cgroup.
+  d. `/kube-proxy`: The kube-proxy is under this cgroup.
+
+4. Protect system processes from memory pressure on the node: The `oom_score_adj` for system processes is set to `-1000`
+
+5. Manage Quality of service for user containers: Three classes of containers are currently supported:
+
+   a. `Guaranteed`: Memory is always guaranteed up to the request. These containers will not OOM unless they exceed their request, or system processes consume too much memory, or the kernel overhead is high and the memory in the node is being fully consumed. 
+
+   b. `Burstable`: Memory is guaranteed up until the usage exceeds user specified `Request`. These containers do not get killed as soon as they exceed their request. Kernel memory overhead and system processes memory usage will impact memory availability for this class of containers. They are more likely to be killed by the kernel OOM killer once they exceed their request. 
+
+   c. `Best-Effort`: Memory is never guaranteed. This class of containers are the first ones to be killed by the kernel whenever there is memory pressure on the node. 
+
+6. Detect OOM kills on containers: Docker tracks OOM kills on its containers. Kubelet exposes this information via [ContainerStatus.State](../pkg/api/v1/types.go#L1033), where the state is set to `ContainerStateTerminated` and `Reason` is set to `OOM Killed`.
+
+#### What are some of the known limitations?
+
+1. The scheduler does not take into account the total usage on the node. Hence, memory can be easily over-committed. Whenever a job is scheduled without resource requests, the node becomes over-committed. 
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. 
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit. A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+
+5. Kubelet does not act on OOM kills. A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer. 
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. Memory usage of system daemons is unrestricted. This results in varying amount of memory availability on the node.
```

Comment:
Today, users are facing issues with kubelet consuming nearly 100% cpu, docker consuming a lot of memory. Because of the fact that the node daemons are unrestricted, it is difficult for the scheduler to avoid overcommitting a node.
I agree with you on not solving this right away because, we do not have control over the node and @dchen1107 Node Spec is required in-order to solve this.